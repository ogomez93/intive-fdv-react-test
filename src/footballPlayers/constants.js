export const NAME = 'footballPlayers';
export const PLAYERS_API = 'https://football-players-b31f2.firebaseio.com/players.json?print=pretty';