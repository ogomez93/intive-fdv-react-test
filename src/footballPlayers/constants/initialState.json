{
  "fetching": true,
  "players": [],
  "nameFilter": "",
  "ageFilter": "",
  "positionFilter": ""
}
