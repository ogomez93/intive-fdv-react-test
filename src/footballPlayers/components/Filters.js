import React from 'react';
import { connect } from 'react-redux';
import { setNameFilterPreview, setPositionFilterPreview, setAgeFilterPreview, applyFilters } from '../actions';
import { POSITIONS } from '../constants';

const positions = POSITIONS.map(position =>
  <option
    value={position}
    key={position}>
    {position}
  </option>
);

const Filters = ({ nameFilterPreview, ageFilterPreview, positionFilterPreview, onNameChange, onPositionChange, onAgeChange, onSearchSubmit }) => {
  let nameInput, positionSelect, ageInput;
  return (
    <form className="my-4 text-xs sm:flex sm:justify-between" onSubmit={ onSearchSubmit }>
      <input
        className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
        onChange={ () => onNameChange(nameInput.value) }
        ref={ node => nameInput = node }
        placeholder="Player Name" />
      <select
        className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white"
        onChange={ () => onPositionChange(positionSelect.value) }
        ref={ node => positionSelect = node }
        placeholder="Position">
        <option value="">Position</option>
        { positions }
      </select>
      <input
        className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
        type="number"
        onChange={ () => onAgeChange(ageInput.value) }
        ref={ node => ageInput = node }
        placeholder="Age"
        min="18"
        max="40" />
      <button
        className="border-2 border-black px-6 py-3 w-full sm:w-auto"
        type="submit">
        Search
      </button>
    </form>
  )
};

const mapStateToProps = (state) => {
  return {
    nameFilterPreview: state.footballPlayers.nameFilterPreview,
    ageFilterPreview: state.footballPlayers.ageFilterPreview,
    positionFilterPreview: state.footballPlayers.positionFilterPreview
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onNameChange: (value) => dispatch(setNameFilterPreview(value)),
    onPositionChange: (value) => dispatch(setPositionFilterPreview(value)),
    onAgeChange: (value) => dispatch(setAgeFilterPreview(value)),
    onSearchSubmit: (event) => {
      event.preventDefault();
      dispatch(applyFilters());
    }
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Filters);