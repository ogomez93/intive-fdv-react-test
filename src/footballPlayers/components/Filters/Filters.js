import React from 'react';
import { POSITIONS } from 'footballPlayers/constants';

const positions = POSITIONS.map(position =>
  <option value={ position } key={ position }>
    { position }
  </option>
);

const Filters = ({
  age,
  name,
  onAgeChange,
  onNameChange,
  onPositionChange,
  onSubmit
}) => (
  <form className="my-4 text-xs sm:flex sm:justify-between" onSubmit={ onSubmit }>
    <input
      name="playerName"
      type='text'
      className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
      onChange={ onNameChange }
      placeholder="Player Name"
      value={ name }
    />

    <select
      name="playerPosition"
      className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white"
      onChange={ onPositionChange }
      placeholder="Position">
      <option value="">Position</option>
      { positions }
    </select>

    <input
      name="playerAge"
      className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
      type="number"
      onChange={ onAgeChange }
      placeholder="Age"
      min="18"
      max="40"
      value={ age }
    />

    <button
      className="border-2 border-black px-6 py-3 w-full sm:w-auto"
      type="submit">
      Search
    </button>
  </form>
);

export default Filters;
