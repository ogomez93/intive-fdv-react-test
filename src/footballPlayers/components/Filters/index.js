import { connect } from 'react-redux';
import { compose, withState, withHandlers } from 'recompose';

import Filters from './Filters';
import { onAgeChange, onNameChange, onPositionChange } from './utils/handlers';
import { applyFilters } from '../../actions/filters';

const mapStateToProps = ({
  footballPlayers: {
    nameFilter,
    ageFilter,
    positionFilter
  }
}) => ({
  nameFilter,
  ageFilter,
  positionFilter
});

const mapDispatchToProps = (dispatch, { name, position, age }) => ({
  onSubmit: event => {
    event.preventDefault();
    dispatch(applyFilters({ name, position, age }));
  }
});

export const enhance = compose(
  withState('name', 'setName', ''),
  withState('position', 'setPosition', ''),
  withState('age', 'setAge', ''),
  withHandlers({
    onNameChange,
    onPositionChange,
    onAgeChange
  }),
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhance(Filters);
