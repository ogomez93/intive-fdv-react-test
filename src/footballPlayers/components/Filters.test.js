import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';

import Filters from './Filters';
import { initialState } from './testConstants/Filters.constants';

const mockStore = configureStore();

const store = mockStore(initialState);

describe('Component: <Filters />', () => {
  const wrapper = shallow(<Filters store={store} />);
  const component = wrapper.dive();
  const componentToJson = toJson(component);

  test('renders the component', () => {
    expect(componentToJson).toMatchSnapshot();
  });

  test('renders player name input', () => {
    const nameInput = componentToJson.children.find((el) => el.props.name === 'playerName');
    expect(nameInput).not.toBeNull();
  });

  test('renders player position select', () => {
    const positionSelect = componentToJson.children.find((el) => el.props.name === 'playerPosition');
    expect(positionSelect).not.toBeNull();
  });

  test('renders player age input', () => {
    const ageInput = componentToJson.children.find((el) => el.props.name === 'playerAge');
    expect(ageInput).not.toBeNull();
  });
});
