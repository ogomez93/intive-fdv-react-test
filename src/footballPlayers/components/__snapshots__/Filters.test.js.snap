// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Filters /> renders the component 1`] = `
<form
  className="my-4 text-xs sm:flex sm:justify-between"
  onSubmit={[Function]}
>
  <input
    className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
    name="playerName"
    onChange={[Function]}
    placeholder="Player Name"
  />
  <select
    className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4 bg-white"
    name="playerPosition"
    onChange={[Function]}
    placeholder="Position"
  >
    <option
      value=""
    >
      Position
    </option>
    <option
      key="Attacking Midfield"
      value="Attacking Midfield"
    >
      Attacking Midfield
    </option>
    <option
      key="Central Midfield"
      value="Central Midfield"
    >
      Central Midfield
    </option>
    <option
      key="Centre-Back"
      value="Centre-Back"
    >
      Centre-Back
    </option>
    <option
      key="Centre-Forward"
      value="Centre-Forward"
    >
      Centre-Forward
    </option>
    <option
      key="Defensive Midfield"
      value="Defensive Midfield"
    >
      Defensive Midfield
    </option>
    <option
      key="Keeper"
      value="Keeper"
    >
      Keeper
    </option>
    <option
      key="Left Midfield"
      value="Left Midfield"
    >
      Left Midfield
    </option>
    <option
      key="Left Wing"
      value="Left Wing"
    >
      Left Wing
    </option>
    <option
      key="Left-Back"
      value="Left-Back"
    >
      Left-Back
    </option>
    <option
      key="Right-Back"
      value="Right-Back"
    >
      Right-Back
    </option>
  </select>
  <input
    className="border-2 border-black p-2 my-1 sm:my-0 w-full sm:w-1/4"
    max="40"
    min="18"
    name="playerAge"
    onChange={[Function]}
    placeholder="Age"
    type="number"
  />
  <button
    className="border-2 border-black px-6 py-3 w-full sm:w-auto"
    type="submit"
  >
    Search
  </button>
</form>
`;
